### PR TITLE
WIP: IEEE 802.15.4 support for nrf52

### PR DESCRIPTION
**Pull request by raiden00pl**
## Summary

- various cosmetic changes related to ieee802154

-  nrf52/nrf52_radio: various changes to support IEEE802154
    - remove read-write logic - this should be handled by radio protocol implementation
    - remove EVENTS and TASKS bit definitions - we can just use a single definition
    - add more radio ops
    - fix frequency configuration
    - fix printf warnings
    - fix radio reset

- arch/nrf52: add initial support for IEEE 802.15.4   
    Supported features:
    
    - frame transmition
    - frame reception and filtering
    - immediate ACK (incoming and outgoing)
    - promiscuous mode
    - delayed transmision
    - radio events trace
    - setting pending bit for all incoming Data Request frames
    - un-slotted CSMA-CA
    
    Work in progres features (some logic is present, but they require more work):
    
    - beacon transmision (periodic transmition works, but requires verification)
    - slotted CSMA-CA
    - GTS
    
    Fetures not implemented:
    
    - enhanced ACK (Enh-ACK)
    - enhanced beacon
    - low power mode
    - advanced features from IEEE 802.15.4e (DSME, TSCH)
    
    Added examples for boards:
    
    - nrf52832-dk: mrf24j40_6lowpan
    - nrf52832-dk: mrf24j40_mac
    - nrf52840-dk: ieee802154_6lowpan
    - nrf52840-dk: ieee802154_mac
    - nrf52840-dongle: ieee802154_mac
    - nrf9160-dk-nrf52: ieee802154_6lowpan
    - nrf9160-dk-nrf52: ieee802154_mac

### CODE REVIEW

**Comment by acassis** [comment]
@raiden00pl are you plan to add 6LoWPAN support to nRF5x ?

**Comment by raiden00pl** [comment]
@acassis Yes, that's the plan. I hope it'll be finished before Nordic releases nrf54 :)

**Comment by raiden00pl** [comment]
Ready for review

File: arch/arm/src/nrf52/CMakeLists.txt [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
 # Unless required by applicable law or agreed to in writing, software
 # distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 # WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
-# License for the specific language governing permissions and limitations
-# under the License.
+# License for the specific language governing permissions and limitations under
+# the License.
 #
-############################################################################
+# ##############################################################################
```
Comment (xiaoxiang781216) [inline arch/arm/src/nrf52/CMakeLists.txt:19]: why change

File: arch/arm/src/nrf52/CMakeLists.txt [hunk 2]
```diff
@@ -1,22 +1,22 @@
...
```
Comment (xiaoxiang781216) [inline arch/arm/src/nrf52/CMakeLists.txt:1]: why add space after first #

File: arch/arm/src/nrf52/nrf52_ieee802154_trace.c [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+#include "nrf52_ieee802154_trace.h"
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* This must match enum radio_trace_type_e */
+
+const static char *g_radio_trace_str[] =
```
Comment (xiaoxiang781216) [inline arch/arm/src/nrf52/nrf52_ieee802154_trace.c:41]: static const

File: arch/arm/src/nrf52/nrf52_ieee802154_trace.c [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  "CSMATRIGGER",
+  "NOCSMATRIGGER",
+  "NOACK",
+  "DROPFRAME",
+  "TXTRYAGAIN",
+};
+
+/* Trace data */
+
+volatile static size_t g_radio_trace_cntr = 0;
```
Comment (xiaoxiang781216) [inline arch/arm/src/nrf52/nrf52_ieee802154_trace.c:80]: static volatile

File: arch/arm/src/nrf52/nrf52_ieee802154_trace.c [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+volatile static struct radio_trace_s
```
Comment (xiaoxiang781216) [inline arch/arm/src/nrf52/nrf52_ieee802154_trace.c:81]: ditto

File: arch/arm/src/nrf52/nrf52_ieee802154_radio.c [hunk 6]
```diff
@@ -0,0 +1,2116 @@
...
+  g_nrf52_radioi8_ackbuf[IEEE802154_ACK_FRAME_SIZE + 1];
+
+#ifdef CONFIG_NRF52_RADIO_IEEE802154_SUPERFRAME
+static uint8_t aligned_data(8)
+  g_nrf52_radioi8_beaconbuf[IEEE802154_MAX_PHY_PACKET_SIZE + 2];
+#endif
+
+/* Radio ops */
+
+struct nrf52_radioi8_radio_ops_s g_radioi8_radio_ops =
```
Comment (xiaoxiang781216) [inline arch/arm/src/nrf52/nrf52_ieee802154_radio.c:231]: add static

File: arch/arm/src/nrf52/nrf52_ieee802154_radio.c [hunk 7]
```diff
@@ -0,0 +1,2116 @@
...
+  .beacon_tx    = nrf52_radioi8_beacon_tx,
+#endif
+  .reset        = nrf52_radioi8_reset,
+  .csma_poll    = nrf52_radioi8_csmapoll,
+  .gts_poll     = nrf52_radioi8_gtspoll
+};
+
+/* Radio interface */
+
+struct nrf52_radioi8_radio_s g_radioi8_radio;
```
Comment (xiaoxiang781216) [inline arch/arm/src/nrf52/nrf52_ieee802154_radio.c:253]: add static

File: arch/arm/src/nrf52/CMakeLists.txt (hunk 1 above)
Comment (raiden00pl) [reply arch/arm/src/nrf52/CMakeLists.txt:19]: cmake-format complains and CI fails. This file was last edited before cmake-format was introduced, so it has not been refreshed yet.

File: arch/arm/src/nrf52/CMakeLists.txt (hunk 2 above)
Comment (raiden00pl) [reply arch/arm/src/nrf52/CMakeLists.txt:1]: cmake-format formtas files this way

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by maxikrie** [comment]
I have done some testing with this on the nrf52840-dk board and I can send UDP packets between two boards. Thank you for your awesome work! @raiden00pl 

However, when I do more heavy duty sending (100 Hz), the server occasionally crashes in the hpwork queue thread. I have increased stack size of the thread with no noticeable improvements. I will investigate this further, but I wanted to provide some feedback with the hope on any leads.

Thanks!

**Pull request merged**